modules/azure/client_factory.go

```diff
@@ -0,0 +1,141 @@
+
+	// GetManagedClustersClientE returns a configured compute client, setup for proper cloud environment use.
+	GetManagedClustersClientE(subscriptionID string) (containerservice.ManagedClustersClient, error)
+}
+
+// multiEnvClientFactory is used to coordinate handing out properly configured Azure SDK clients
+// that are properly setup for use with Public or Sovereign clouds (depending on configuration)
+type multiEnvClientFactory struct{}
+
+// NewClientFactory returns a new multi-environment client factory
+func NewClientFactory() ClientFactory {
+	return &multiEnvClientFactory{}
+}
+
+// GetVirtualMachinesClientE returns a configured compute client, setup for proper cloud environment use.
+func (factory *multiEnvClientFactory) GetVirtualMachinesClientE(subscriptionID string) (compute.VirtualMachinesClient, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return compute.VirtualMachinesClient{}, err
+	}
+
+	// Lookup environment URI
+	baseURI, err := factory.getEnvironmentBaseURI()
+	if err != nil {
+		return compute.VirtualMachinesClient{}, err
+	}
+
+	// Create a VM client and return
+	vmClient := compute.NewVirtualMachinesClientWithBaseURI(baseURI, subscriptionID)
+	return vmClient, nil
+}
+
+// GetSubscriptionClientE returns a configured compute client, setup for proper cloud environment use.
+func (factory *multiEnvClientFactory) GetSubscriptionClientE() (subscriptions.Client, error) {
+	// Lookup environment URI
+	baseURI, err := factory.getEnvironmentBaseURI()
+	if err != nil {
+		return subscriptions.Client{}, err
+	}
+
+	// Create a Subscription client
+	client := subscriptions.NewClientWithBaseURI(baseURI)
+	return client, nil
+}
+
+// GetManagedClustersClientE returns a configured compute client, setup for proper cloud environment use.
+func (factory *multiEnvClientFactory) GetManagedClustersClientE(subscriptionID string) (containerservice.ManagedClustersClient, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return containerservice.ManagedClustersClient{}, err
+	}
+
+	// Lookup environment URI
+	baseURI, err := factory.getEnvironmentBaseURI()
+	if err != nil {
+		return containerservice.ManagedClustersClient{}, err
+	}
+
+	client := containerservice.NewManagedClustersClientWithBaseURI(baseURI, subscriptionID)
+	return client, nil
+}
+
+// getDefaultEnvironmentName returns either a configured Azure environment name, or the public default
+func (factory *multiEnvClientFactory) getDefaultEnvironmentName() string {
+	envName, exists := os.LookupEnv(AzureEnvironmentEnvName)
+
+	if !exists || envName == "" {
+		envName = autorest.PublicCloud.Name
+	}
+
+	return envName
+}
+
+// getEnvironmentBaseUri returns the ARM management URI for the configured Azure environment.
+func (factory *multiEnvClientFactory) getEnvironmentBaseURI() (string, error) {
```

Comment:
should these methods have an E on end of name?

Reply:
Great catch - yes they should to follow conventions in the repo. Will adjust.